docs/pallas/tpu/pipelining.md

```diff
@@ -0,0 +1,451 @@
+  return pl.pallas_call(
+      naive_sum_kernel,
+      grid=grid,
+      # None in `block_shape` means we pick a size of 1 and squeeze it away
+      in_specs=[pl.BlockSpec(lambda i: (i, 0, 0), (None, *out_shape))],
+      out_specs=pl.BlockSpec(lambda i: (0, 0), out_shape),
+      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
+      )(x)
+naive_sum(x)
+```
+
++++ {"id": "Kv9qJYJY4jbK"}
+
+Notice how we've set up the `BlockSpec`s: we're loading the entirety of the `(512, 512)` dimension into VMEM (no pipelining there) but selecting the `i`-th dimension of `x` each iteration in the `index_map`. We are using a `None` for that dimension in the block shape, which indicates that we are selecting a singleton dimension from `x` that we would like squeeze away in the kernel. Therefore, `x_ref` is `(512, 512)`-shaped in VMEM as well.
+
+`out_spec` uses `lambda i: (0, 0)` as its `index_map`, indicating that `o_ref` is unchanged over the course of the pipeline. This means that we can update its value each iteration by reading from and writing to it. Or can it? Actually there is one catch: *`o_ref` is initially garbage*, meaning we'll be accumulating into garbage. This will result in the overall function outputting the incorrect value!
+
+Therefore, **whenever we do a reduction in a kernel, we need to make sure to initialize the `Ref` that is storing the reduced value**. We can accomplish this using `pl.when` (which executes a function conditionally) and `pl.program_id` (which queries which iteration of the grid inside of a kernel).
+
+```{code-cell}
+---
+executionInfo:
+  elapsed: 91
+  status: ok
+  timestamp: 1708910182754
+  user:
+    displayName: Sharad Vikram
+    userId: '16113763449895047367'
+  user_tz: 480
+id: JXN2RthX5cSw
+outputId: 195df19b-a889-479b-95b6-1fb7281f1518
+---
+def sum_kernel(x_ref, o_ref):
+  @pl.when(pl.program_id(0) == 0)
+  def _():
+    o_ref[...] = jnp.zeros_like(o_ref)
+
+  o_ref[...] += x_ref[...]
+
+def sum(x: jax.Array) -> jax.Array:
+  grid, *out_shape = x.shape
+  return pl.pallas_call(
+      sum_kernel,
+      grid=grid,
+      # None in `block_shape` means we pick a size of 1 and squeeze it away
+      in_specs=[pl.BlockSpec(lambda i: (i, 0, 0), (None, *out_shape))],
+      out_specs=pl.BlockSpec(lambda i: (0, 0), out_shape),
+      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
+      )(x)
+sum(x)
+```
+
++++ {"id": "2828qXBI5ksZ"}
+
+This `sum` function now outputs the correct values!
+
+One last thing to note about reductions in Pallas are that **they must be done in the minormost (rightmost) dimensions of our grid**. This is because the pipeline that Pallas generates using the `BlockSpec`s, `grid` and kernel function *does not read outputs back from HBM*. Once you've written an output value back to HBM you cannot revisit it. Therefore, you cannot do a reduction across a grid dimension that has any revisiting and therefore all reductions need to happen in the rightmost dimensions.
```

Comment:
You say to reduce on the rightmost dimension but above you are reducing on the leftmost (major) dimension right?

Maybe the example should be changed?

Reply:
The `grid` is only one dimensional so technically we are reducing over its minormost dimension (not the minormost dimension of the inputs/outputs, it's the grid that matters). I'll add clarification.

Reply:
Ahh, makes sense. Thanks!